lib/proxy/conn.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/trace"
+	"google.golang.org/grpc"
+)
+
+// Stream is a common interface for grpc client and server streams.
+type Stream interface {
+	Context() context.Context
+	Send(*proto.Frame) error
+	Recv() (*proto.Frame, error)
+}
+
+// streamConn wraps a grpc stream with a net.streamConn interface.
+type streamConn struct {
+	stream Stream
+
+	wLock  sync.Mutex
+	rLock  sync.Mutex
+	rBytes []byte
+	rOff   int
+
+	src net.Addr
+	dst net.Addr
+}
+
+// newStreamConn creates a new streamConn.
+func newStreamConn(stream Stream, src net.Addr, dst net.Addr) *streamConn {
+	return &streamConn{
+		stream: stream,
+		src:    src,
+		dst:    dst,
+	}
+}
+
+// Read reads data reveived over the grpc stream.
+func (c *streamConn) Read(b []byte) (n int, err error) {
+	c.rLock.Lock()
+	defer c.rLock.Unlock()
+
+	if c.rOff == 0 {
+		frame, err := c.stream.Recv()
+		if err != nil {
+			return 0, trace.ConnectionProblem(err, "failed to receive on stream")
+		}
+
+		data := frame.GetData()
+		if data == nil {
+			return 0, trace.ConnectionProblem(trace.Errorf("failed to receive on stream"), "invalid data frame")
+		}
+
+		c.rBytes = data.Bytes
+	}
+
+	copy(b, c.rBytes[c.rOff:])
+
+	// Reset the read offset if all the data has been read.
+	if (len(c.rBytes) - c.rOff) <= len(b) {
+		n := len(c.rBytes) - c.rOff
+		c.rOff = 0
+		return n, nil
+	}
+
+	c.rOff += len(b)
+	return len(b), nil
+}
+
+// Write sends data over the grpc stream.
+func (c *streamConn) Write(b []byte) (int, error) {
+	c.wLock.Lock()
+	defer c.wLock.Unlock()
+
+	err := c.stream.Send(&proto.Frame{Message: &proto.Frame_Data{Data: &proto.Data{
+		Bytes: b,
```

Comment:
What if `b` is very big and we exceed the grpc message limit? I think we ought to chunk this out.

Reply:
Chunking added here [c3dda48](https://github.com/gravitational/teleport/pull/10223/commits/c3dda486df5ed4fba966ee0273f5b2c01dbd02ce)